src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -214,6 +217,82 @@ public ValueTask<FlushResult> WriteResponseTrailers(int streamId, HttpResponseTr
             }
         }
 
+        private static int EncodeStatusCode(int statusCode, Span<byte> buffer)
+        {
+            switch (statusCode)
+            {
+                // Status codes which exist in the HTTP/2 StaticTable.
+                case 200:
+                case 204:
+                case 206:
+                case 304:
+                case 400:
+                case 404:
+                case 500:
+                    buffer[0] = (byte)(0x80 | H2StaticTable.StatusIndex[statusCode]);
+                    return 1;
+                default:
+                    // Send as Literal Header Field Without Indexing - Indexed Name
+                    buffer[0] = 0x08;
+
+                    ReadOnlySpan<byte> statusBytes = System.Net.Http.HPack.StatusCodes.ToStatusBytes(statusCode);
+                    buffer[1] = (byte)statusBytes.Length;
+                    statusBytes.CopyTo(buffer.Slice(2));
+
+                    return 2 + statusBytes.Length;
+            }
+        }
+
+        private static bool EncodeHeaders(int statusCode, Http2HeadersEnumerator headersEnumerator, Span<byte> buffer, out int length)
+        {
+            var statusCodeLength = EncodeStatusCode(statusCode, buffer);
+            var done = EncodeHeaders(headersEnumerator, buffer.Slice(statusCodeLength), throwIfNoneEncoded: false, out var headersLength);
+            length = statusCodeLength + headersLength;
+
+            return done;
+        }
+
+        private static bool EncodeHeaders(Http2HeadersEnumerator headersEnumerator, Span<byte> buffer, bool throwIfNoneEncoded, out int length)
+        {
+            var currentLength = 0;
+            do
+            {
+                if (!EncodeHeader(headersEnumerator.KnownHeaderType, headersEnumerator.Current.Key, headersEnumerator.Current.Value, buffer.Slice(currentLength), out int headerLength))
+                {
+                    // The the header wasn't written and no headers have been written then the header is too large.
+                    // Throw an error to avoid an infinite loop of attempting to write large header.
+                    if (currentLength == 0 && throwIfNoneEncoded)
+                    {
+                        throw new HPackEncodingException(SR.net_http_hpack_encode_failure);
+                    }
+
+                    length = currentLength;
+                    return false;
+                }
+
+                currentLength += headerLength;
+            }
+            while (headersEnumerator.MoveNext());
+
+            length = currentLength;
+
+            return true;
+        }
+
+        private static bool EncodeHeader(KnownHeaderType knownHeaderType, string name, string value, Span<byte> buffer, out int length)
```

Comment:
I had envisioned this method and the enum being in the shared code, but I can understand why you would want to stop here. The client has almost this exact code for request headers but it will take some more work to deduplicate.

Reply:
We're at least now using the same static helper methods.

I'm now sure where the client would use a method like EncodeHeader. Many of their headers are written explicitly, and their unknown headers are written here - https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L1021

I'm not sure how their header buffer handles splitting content across multiple HEADER frames.

Reply:
I meant that translations like HPackHeaderMapping.GetResponseHeaderStaticTableId would be done inside the HPack implementation.